promql/engine.go

```diff
@@ -2802,7 +2803,15 @@ func vectorElemBinop(op parser.ItemType, lhs, rhs float64, hlhs, hrhs *histogram
 			}
 			return 0, res.Compact(0), true, nil
 		}
-		return lhs + rhs, nil, true, nil
+		if hlhs == nil && hrhs == nil {
+			return lhs + rhs, nil, true, nil
+		}
+		lhsTyp := "float"
+		rhsTyp := "histogram"
+		if hlhs != nil {
+			lhsTyp, rhsTyp = rhsTyp, lhsTyp
+		}
+		return 0, nil, false, annotations.NewIncompatibleTypesInBinOpInfo(lhsTyp, "+", rhsTyp, pos)
```

Comment:
Maybe this is a bit too tricksy for the effect. How about the following (fewer lines, but more characters, overall more explicit):

```suggestion
		if hlhs != nil {
			return 0, nil, false, annotations.NewIncompatibleTypesInBinOpInfo("histogram", "+", "float", pos)
		}
		return 0, nil, false, annotations.NewIncompatibleTypesInBinOpInfo("float", "+", "histogram", pos)
```

Reply:
Yes, you are right. It will be more explicit in that way. I'll change it.